MDEV-26474: Fix mysql_setpermission hostname logic

mkulik-rh opened the pull request:
<!--
If you've already identified a https://jira.mariadb.org/ issue
that seems to track this bug/feature, please add its number below.
-->
- [ ] *The Jira issue number for this PR is: MDEV-_____*


## Description
Hi,

It seems that there is some kind of logic confusion in mysql set_permission perl script. By default all connections parameters are read from configuration file of mariadb:

https://github.com/MariaDB/server/blob/10.7/scripts/mysql_setpermission.sh#L61

Fresh installation of mariadb on Fedora does not contain any connection information in configuration file. It falls back automatically to localhost DB. Nothing wrong with that.

Trying to execute mysql_setpermission with this configuration will fail:

```
mysql_setpermission fails to connect to local mariadb, claiming that 
DBI connect(';host=localhost;port=3306','newuser',...) failed: Connection error: port cannot be specified when host is localhost or embedded at /usr/bin/mysql_setpermission line 116.
```

I don't think that's correct behavior. It should be possible to use it on default installation without additional configuration.

Here is what's going on. Reading host from my.cnf file will result in no information, so script will fallback to 'localhost' string:

https://github.com/MariaDB/server/blob/10.7/scripts/mysql_setpermission.sh#L76

Seems like a good idea but unfortunately DBD module in perl will treat is as socket connection:

The hostname, if not specified or specified as '' or 'localhost', will default to a MySQL server running on the local machine using the default for the UNIX socket. To connect to a MySQL server on the local machine via TCP, you must specify the loopback IP address (127.0.0.1) as the host.

Changing it to 127.0.0.1 will solve the issue, although I can see another confusion. User might specify hostname in my.cnf file as localhost fot TCP connection and that will be totally fine for mariasb but not for this module, so another condition is required to catch this exception.

## How can this PR be tested?
```
# systemctl start mariadb
# mysql
mysql> CREATE USER 'newuser'@'localhost' IDENTIFIED BY 'password';
mysql> GRANT ALL PRIVILEGES ON *.*  TO 'newuser'@'localhost' WITH GRANT OPTION;
mysql> FLUSH PRIVILEGES;
# mysql_setpermission  --user newuser
```

<!--
Tick one of the following boxes [x] to help us understand
if the base branch for the PR is correct
-->
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [x] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=1896) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=1896) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=1896) it.</sub>

grooverdan commented:
Hi,

Thanks for the pull request, your first I see. Welcome!

Good description of the problem. I don't think correcting the hostname is right approach.

The correct fix I think is on line 112 when the DSN is formed, if the hostname is `localhost` don't append `;port=`. This preserves the compatibility as much as possible for existing users who would otherwise get an error message.

If you want to change that, test it, I'm happy to take a fix into the 10.2 branch. To do that recreate your fix off the 10.2 branch. `git push --force` to the same `setperm_fix` branch. And in the github issue here edit the target branch to 10.2,

While you are correcting this, can I ask for a second fix (a second commit on the same branch is ok) -> remove the `FLUSH PRIVILEGES` line. This hasn't been needed for very many years.

mkulik-rh commented:
Hi,
Thanks.

Yours solution also works correctly. I changed base branch to 10.2. Code for 10.7 will look a little bit different during merge conflict:

```perl
if ($opt_socket and -S $opt_socket)
{
    $dsn .= "${prefix}_socket=$opt_socket";
}
else
{
  $dsn .= "host=$sqlhost";
  if ($sqlhost ne "localhost") {
    $dsn .= ";port=$opt_port";
  }
}

# make the connection to MariaDB
$dbh= DBI->connect($dsn,$opt_user,$opt_password, { RaiseError => 1, PrintError => 0}) ||
  die("Can't make a connection to the MariaDB server.\n The error: $DBI::errstr");

``` 
I also added commit that removes FLUSH PRIVILEGES from mysql_setpermission.

grooverdan commented:
Looks good. Thanks for the test and 10.7 alternate.

Can you just confirm contribution under the BSD-new license please?

mkulik-rh commented:
I signed it using MariaDB Contributor Agreement and I cannot change it to New BSD License (3-clause).

grooverdan commented:
All good. Thank you.

grooverdan commented:
merged in 10.7 implementation too 6a987149a8151bffdf6ef9f21989ae894c9e18f5
